src/Evolution/Systems/ScalarTensor/Sources/ScalarSource.hpp

```diff
@@ -0,0 +1,90 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "DataStructures/DataBox/Tag.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Tags.hpp"
+#include "Evolution/Systems/ScalarTensor/Sources/Tags.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace ScalarTensor {
+
+/*!
+ * \brief Add in the source term to the \f$\Pi\f$
+ * evolved variable of the ::CurvedScalarWave system.
+ *
+ * \details The only source term in the wave equation
+ * \f[
+ *  \Box \Psi = \mathcal{S} ~,
+ *  \f]
+ *
+ * is in the equation for \f$\Pi\f$:
+ * \f[
+ *  \partial_t \Pi + \text{\{spatial derivative
+ * terms\}} = \alpha \mathcal{S}
+ * ~,
+ * \f]
+ *
+ * where \f$\mathcal{S}\f$ is the source term (e. g. in the Klein-Gordon
+ * equation, the source term is the derivative of the scalar potential
+ * \f$\mathcal{S} \equiv \partial V / \partial \Psi \f$.)
+ *
+ * (Note that this function takes as argument the source term.)
+ *
+ * This function adds that contribution to the existing value of `dt_pi_scalar`.
+ * The wave equation terms in the scalar equation should be computed before
+ * passing the `dt_pi_scalar` to this function for updating.
+ *
+ * \see `CurvedScalarWave::TimeDerivative` for details about the source-less
+ * part of the time derivative calculation.
+ */
+void add_scalar_source_to_dt_pi_scalar(
+    gsl::not_null<Scalar<DataVector>*> dt_pi_scalar,
+    const Scalar<DataVector>& scalar_source, const Scalar<DataVector>& lapse);
+
+/*!
+ * \brief Computes the source term given by the mass of the scalar.
+ *
+ * \details For a scalar field with mass parameter \f$ m_\Psi \f$,
+ * the wave equation takes the form
+ * \f[
+ *   \Box \Psi = \mathcal{S} ~,
+ * \f]
+ *
+ * where the source is given by
+ * \f[
+ *   \mathcal{S} \equiv m^2_\Psi ~.
+ * \f]
+ *
+ * Here the mass parameter value is an option that needs to be specified in the
+ * input file.
+ *
+ * \see `ScalarTensor::Tags::ScalarMass` for details about the mass.
+ */
+void compute_mass_source(gsl::not_null<Scalar<DataVector>*> scalar_source,
+                           const Scalar<DataVector>& psi,
+                           const double mass_psi);
+
+namespace Tags {
+
+/*!
+ * \brief Compute tag for the scalar source.
+ *
+ * \details Compute the scalar source from data box items using
+ * `compute_mass_source`.
+ */
+struct ScalarSourceCompute : ScalarSource, db::ComputeTag {
```

Comment:
My recommendation is to avoid compute items if they're part of your algorithm. It's a lot easier to understand if algorithms are implemented in standard C++ functions rather than through a hierarchy of compute items. Computing things in a function can also be optimized better, reusing memory, etc. Compute items are mostly useful for observing derived quantities. So if you want to use a compute item here or not depends on your use case.

Reply:
Thanks for the recommendation. This is actually an important point mainly for the following TimeDerivative PR https://github.com/sxs-collaboration/spectre/pull/5212 . I wrote this system using a quite few compute tags for some of the gr quantities and empirically it seems not to be that much slower than gh. I think this can be refined/optimized in a second pass later as it might need a little bit of care to implement.

Reply:
Ok. Also keep in mind that compute items take up memory throughout the full simulation. If you need to re-compute a quantity in every time step and don't need to observe it (i.e. temporary / intermediate quantities), then prefer computing it directly in the time derivative function and reuse some of the temporary memory.